lib/maths/CModel.cc

```diff
@@ -130,6 +130,17 @@ CModelAddSamplesParams::priorWeights() const {
     return *m_PriorWeights;
 }
 
+CModelAddSamplesParams&
+CModelAddSamplesParams::onModelChange(const TModelChangeCallback& modelChangeCallback) {
+    m_ModelChangeCallback = &modelChangeCallback;
+    return *this;
+}
+
+const CModelAddSamplesParams::TModelChangeCallback&
+CModelAddSamplesParams::onModelChange() const {
+    return *m_ModelChangeCallback;
```

Comment:
This line will cause a fatal crash if either:

1. `m_ModelChangeCallback` is `nullptr`
2. `m_ModelChangeCallback` is not `nullptr` but the `TModelChangeCallback` it points at is an empty `std::function`

It would be much safer if:

1. `m_ModelChangeCallback` was held by value rather than by pointer
2. `m_ModelChangeCallback` defaulted to a no-op function if unset

I guess the reason you didn't do this originally is that this would cause construction and destruction of may `std::function` objects since we would have to construct and destruct for every piece of input data.  In that case I think this method should at least be defensive, i.e.:

```
    if (m_ModelChangeCallback != nullptr && *m_ModelChangeCallback != nullptr) {
        return *m_ModelChangeCallback;
    } else {
        LOG_ERROR(<< "m_ModelChangeCallback unexpectedly not set");
        return reference_to_a_static_TModelChangeCallback_object_that_is_a_noop;
    }
```

Although it shouldn't happen, these runtime crashes caused by null pointers or empty functions are incredibly difficult to track down if they happen in production.

Reply:
As above, I prefer the option of this not being null and defaulting to a no-op if it isn't overridden in the class.